### PR TITLE
Relax `on_train_batch_*` hook check with `dataloader_iter` to a warning

### PR DESCRIPTION
**Pull request by erhoo82**
## What does this PR do?

This changes the check of using `on_train_batch_start` and `on_train_batch_end` hooks with `dataloader_iter` from `assert` to `warn`. 

When these hooks are used with `dataloader_iter`, the `batch_idx` argument would not match with the training index depending on how often the batch iteration is called per training step. So, this PR intends to raise `warning` on such behavior and let users use the training batch hooks for extensive use of `dataloader_iter`.

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: src/pytorch_lightning/trainer/configuration_validator.py [hunk 1]
```diff
@@ -173,15 +173,17 @@ def __check_training_step_requires_dataloader_iter(model: "pl.LightningModule")
     if is_param_in_hook_signature(training_step_fx, "dataloader_iter", explicit=True):
 
         if is_overridden("on_train_batch_start", model):
-            raise MisconfigurationException(
-                "The model hook `on_train_batch_start` is not compatible with "
-                "taking a `dataloader_iter` argument in your `training_step`."
+            rank_zero_warn(
+                "`batch_idx` arguement in `on_train_batch_start` hook may "
```
Comment (carmocca) [inline src/pytorch_lightning/trainer/configuration_validator.py:177]:
```suggestion
                "`batch_idx` argument in `on_train_batch_start` hook may "
```

File: src/pytorch_lightning/trainer/configuration_validator.py [hunk 2]
```diff
@@ -173,15 +173,17 @@ def __check_training_step_requires_dataloader_iter(model: "pl.LightningModule")
...
+                "not match with the actual batch index when taking `dataloader_iter` "
+                "argument in your `training_step`."
             )
 
         if is_overridden("on_train_batch_end", model):
-            raise MisconfigurationException(
-                "The model hook `on_train_batch_end` is not compatible with "
-                "taking a `dataloader_iter` argument in your `training_step`."
+            rank_zero_warn(
+                "`batch_idx` arguement in `on_train_batch_end` hook may "
```
Comment (carmocca) [inline src/pytorch_lightning/trainer/configuration_validator.py:184]:
```suggestion
                "`batch_idx` argument in `on_train_batch_end` hook may "
```

File: src/pytorch_lightning/trainer/configuration_validator.py [hunk 3]
```diff
@@ -173,15 +173,17 @@ def __check_training_step_requires_dataloader_iter(model: "pl.LightningModule")
     if is_param_in_hook_signature(training_step_fx, "dataloader_iter", explicit=True):
```
Comment (carmocca) [inline src/pytorch_lightning/trainer/configuration_validator.py:173]: I think it would be better to not call these 2 hooks under this condition. Then it is up to the user to manually call them if it's desired.

File: src/pytorch_lightning/trainer/configuration_validator.py (hunk 1 above)
Comment (awaelchli) [inline src/pytorch_lightning/trainer/configuration_validator.py:177]: Let's make a full sentence out of it:
```suggestion
                "The `batch_idx` argument in `on_train_batch_start` hook may "
```

**Review by awaelchli — APPROVED** [review]

File: src/pytorch_lightning/trainer/configuration_validator.py (hunk 1 above)
Comment (erhoo82) [reply src/pytorch_lightning/trainer/configuration_validator.py]: Done

File: src/pytorch_lightning/trainer/configuration_validator.py (hunk 3 above)
Comment (erhoo82) [reply src/pytorch_lightning/trainer/configuration_validator.py:173]: I am also fine with this idea.

**Comment by erhoo82** [comment]
@carmocca 
Please help review and merge.

File: src/pytorch_lightning/trainer/configuration_validator.py (hunk 3 above)
Comment (carmocca) [reply src/pytorch_lightning/trainer/configuration_validator.py:173]: I looked further into this. This would be possible but we still need to call it automatically for the strategies and callbacks. Most importantly, some of our builtin callbacks rely on this so it would be weird to not do the same for the LightningModule. Resolving

**Review by carmocca — APPROVED** [review]
